### ISSUE TITLE
Graphite-web Upgrade error

### ISSUE THREAD

**Issue by njchandu** [graphite-project/graphite-web#2391]
I'm trying to upgrade graphite-web to 1.1.4 from 0.9.15. I'm getting the below error when trying to execute command: `django-admin.py runserver --pythonpath /opt/graphite/webapp --settings graphite.settings 0.0.0.0:8080`

```
/opt/graphite/webapp/graphite/settings.py:332: UserWarning: SECRET_KEY is set to an unsafe default. This should be set in local_settings.py for better security
  warn('SECRET_KEY is set to an unsafe default. This should be set in local_settings.py for better security')
/opt/graphite/webapp/graphite/settings.py:332: UserWarning: SECRET_KEY is set to an unsafe default. This should be set in local_settings.py for better security
  warn('SECRET_KEY is set to an unsafe default. This should be set in local_settings.py for better security')
Performing system checks...

System check identified no issues (0 silenced).

You have 20 unapplied migration(s). Your project may not work properly until you apply the migrations for app(s): account, admin, auth, contenttypes, dashboard, events, sessions, tagging, tags, url_shortener.
Run 'python manage.py migrate' to apply them.

December 07, 2018 - 20:51:41
Django version 1.11, using settings 'graphite.settings'
Starting development server at http://0.0.0.0:8080/
Quit the server with CONTROL-C.
/opt/graphite/webapp/graphite/settings.py:332: UserWarning: SECRET_KEY is set to an unsafe default. This should be set in local_settings.py for better security
  warn('SECRET_KEY is set to an unsafe default. This should be set in local_settings.py for better security')
Unhandled exception in thread started by <function wrapper at 0x7f1b85a6c758>
Traceback (most recent call last):
  File "/usr/local/lib64/python2.7/site-packages/django/utils/autoreload.py", line 227, in wrapper
    fn(*args, **kwargs)
  File "/usr/local/lib64/python2.7/site-packages/django/core/management/commands/runserver.py", line 117, in inner_run
    autoreload.raise_last_exception()
  File "/usr/local/lib64/python2.7/site-packages/django/utils/autoreload.py", line 250, in raise_last_exception
    six.reraise(*_exception)
  File "/usr/local/lib64/python2.7/site-packages/django/utils/autoreload.py", line 227, in wrapper
    fn(*args, **kwargs)
  File "/usr/local/lib64/python2.7/site-packages/django/__init__.py", line 27, in setup
    apps.populate(settings.INSTALLED_APPS)
  File "/usr/local/lib64/python2.7/site-packages/django/apps/registry.py", line 108, in populate
    app_config.import_models()
  File "/usr/local/lib64/python2.7/site-packages/django/apps/config.py", line 202, in import_models
    self.models_module = import_module(models_module_name)
  File "/usr/lib64/python2.7/importlib/__init__.py", line 37, in import_module
    __import__(name)
  File "/opt/graphite/webapp/graphite/events/models.py", line 4, in <module>
    from tagging.models import Tag
  File "/usr/local/lib/python2.7/site-packages/tagging/models.py", line 10, in <module>
    from django.contrib.contenttypes.generic import GenericForeignKey
ImportError: No module named generic

```

Command `pip show django whipser carbon graphite-web` results in

```
Name: Django
Version: 1.11
Summary: A high-level Python Web framework that encourages rapid development and clean, pragmatic design.
Home-page: https://www.djangoproject.com/
Author: Django Software Foundation
Author-email: foundation@djangoproject.com
License: BSD
Location: /usr/local/lib64/python2.7/site-packages
Requires: pytz
Required-by:
```

Any help appreciated, thanks.

**Comment by piotr1212** [comment]
Could you check if your django-tagging module is out of date? I suspect it needs to be updated.

**Comment by njchandu** [comment]
I updated `django-tagging` to `0.4.5`. I'm still getting the same error.

**Comment by njchandu** [comment]
I bumped it to 0.4.6.

```[ec2-user@ip-172-17-71-91 ~]$ pip install django-tagging
Requirement already satisfied: django-tagging in /usr/local/lib/python2.7/site-packages (0.4.6)
```

Is there another way to check what version of django-tagging I have?

**Comment by piotr1212** [comment]
check `/usr/local/lib/python2.7/site-packages/tagging/models.py` and see what is being imported, if it is `from django.contrib.contenttypes.generic import GenericForeignKey` then there is something off. It mus not contain the `.generic` part.

**Comment by njchandu** [comment]
```
# For the ImportError: No module named fields. http://www.reemachugani.com/blog/2015/feb/14/no-module-named-fields-importing-GenericForeignKey/
# In the official django documentation for ContentTypes, the import statement used for GenericForeignKey is given as - `from django.contrib.contenttypes.fields import GenericForeignKey`
# This will raise an ImportError. Replace it with the following - `from django.contrib.contenttypes.generic import GenericForeignKey`
cd /usr/local/lib/python2.7/site-packages
find . -type f -exec sed -i 's/django.contrib.contenttypes.fields/django.contrib.contenttypes.generic/' {} +
```
Found the culprit, this was replacing the `fields` with `generic`.
The error is gone, but graphite doesn't start up

I'm following this doc: https://www.digitalocean.com/community/tutorials/how-to-keep-effective-historical-logs-with-graphite-carbon-and-collectd-on-centos-7 to see what am I doing wrong.

**Comment by deniszh** [comment]
Documentation looks sane. What do you mean when saying "graphite doesn't start up"? graphite-web doesn't start? apache doesn't start? carbon doesn't start? what's in the logs?

**Comment by njchandu** [comment]
As in the dashboard UI doesn't load.

```
django-admin.py runserver --pythonpath /opt/graphite/webapp --settings graphite.settings 0.0.0.0:8080
/opt/graphite/webapp/graphite/settings.py:332: UserWarning: SECRET_KEY is set to an unsafe default. This should be set in local_settings.py for better security
  warn('SECRET_KEY is set to an unsafe default. This should be set in local_settings.py for better security')
/opt/graphite/webapp/graphite/settings.py:332: UserWarning: SECRET_KEY is set to an unsafe default. This should be set in local_settings.py for better security
  warn('SECRET_KEY is set to an unsafe default. This should be set in local_settings.py for better security')
Performing system checks...

System check identified no issues (0 silenced).

You have 21 unapplied migration(s). Your project may not work properly until you apply the migrations for app(s): account, admin, auth, contenttypes, dashboard, events, sessions, tagging, tags, url_shortener.
Run 'python manage.py migrate' to apply them.

December 09, 2018 - 18:35:56
Django version 1.11, using settings 'graphite.settings'
Starting development server at http://0.0.0.0:8080/
Quit the server with CONTROL-C.
```

This is what I see in UI.
```
This page isn’t working dashboard-stress-internal.devaws.net didn’t send any data.
ERR_EMPTY_RESPONSE
```

**Comment by DanCech** [comment]
Did you try setting the SECRET_KEY and running the migrations as instructed in the log output above?

**Comment by njchandu** [comment]
@DanCech I'll try that.

**Comment by njchandu** [comment]
@DanCech @piotr1212 
I tried a few combination of versions for a couple of days. I was not able to solve it, here are the details.
Versions:
django: 1.11.17
twisted: 18.7.0
graphite-web: 1.1.4
whisper: 1.1.4
carbon: 1.1.4
go-carbon: 0.13.0
https://github.com/lomik/go-carbon/releases/download/v0.13.0/go-carbon-0.13.0-1.x86_64.rpm
PYTHONPATH="/opt/graphite/lib/:/opt/graphite/webapp/"

This is the error I'm getting now.
```
[ec2-user@ip ~]$ ./dashboard.sh
starting carbon
2018/12/11 18:30:40 [WARNING] `common.log-level` and `common.logfile` is DEPRICATED. Use `logging` config section
starting graphite-web
starting dashboard
nohup: redirecting stderr to stdout
Starting dxedge-dashboard at port 9000...                             ERROR [Returned: 7]
ERROR [no response from server, timeout]
```
dashboard.sh
```
#!/bin/bash

source /home/ec2-user/.bashrc
export PATH=$PATH:/usr/local/bin

echo "starting carbon"
go-carbon -config /usr/local/etc/go-carbon.conf -daemon

echo "starting graphite-web"
nohup /opt/graphite/bin/run-graphite-devel-server.py /opt/graphite > graphite.out &

cd {{ dashboard_home }}
echo "starting dashboard"
mkdir logs
bin/service.sh start {{ dashboard_port }}
```

The same set up works fine for below versions.
Versions:
django: 1.6.7
twisted: 13.1.0
graphite-web: 0.9.15
whisper: 0.9.15
carbon: 0.9.15
go-carbon: 0.9.1
https://github.com/lomik/go-carbon/releases/download/v0.9.1/go-carbon-0.9.1-1.el6.x86_64.rpm

I'm using ansible for automation and I have removed this bit of code and there are no script changes apart from the version changes.
```
# Fix problems with Python imports.
# Because `defaults` is deprecated. See http://stackoverflow.com/questions/21058177/no-module-named-defaults-django-simple-friends
cd /opt/graphite
find . -type f -exec sed -i 's/django.conf.urls.defaults/django.conf.urls/' {} +
# For the ImportError: No module named fields. http://www.reemachugani.com/blog/2015/feb/14/no-module-named-fields-importing-GenericForeignKey/
# In the official django documentation for ContentTypes, the import statement used for GenericForeignKey is given as - `from django.contrib.contenttypes.fields import GenericForeignKey`
# This will raise an ImportError. Replace it with the following - `from django.contrib.contenttypes.generic import GenericForeignKey`
cd /usr/local/lib/python2.7/site-packages
find . -type f -exec sed -i 's/django.contrib.contenttypes.fields/django.contrib.contenttypes.generic/' {} +
```

Would appreciate any help on this.

**Comment by deniszh** [comment]
@njchandu : unfortunately, scripts above (e.g. `bin/service.sh start` ) are not part of Graphite and we can't say anything about them or Ansible automation. 
But from information above I can say that you didn't apply migrations, please do that using something like 
``` 
cd /opt/graphite
sudo PYTHONPATH=/opt/graphite/webapp/ django-admin.py syncdb --settings=graphite.settings
```
or
```
cd /opt/graphite
sudo PYTHONPATH=/opt/graphite/webapp/ django-admin.py migrate --settings=graphite.settings --run-syncdb
```

**Comment by njchandu** [comment]
I tried `django-admin.py migrate --settings=graphite.settings --run-syncdb`
That seemed to go through successfully. Here is the output.
```
/opt/graphite/webapp/graphite/settings.py:332: UserWarning: SECRET_KEY is set to an unsafe default. This should be set in local_settings.py for better security
  warn('SECRET_KEY is set to an unsafe default. This should be set in local_settings.py for better security')
Operations to perform:
  Synchronize unmigrated apps: browser, composer, functions, metrics, render, staticfiles, whitelist
  Apply all migrations: account, admin, auth, contenttypes, dashboard, events, sessions, tagging, tags, url_shortener
Synchronizing apps without migrations:
  Creating tables...
    Running deferred SQL...
Running migrations:
  Applying contenttypes.0001_initial... OK
  Applying auth.0001_initial... OK
  Applying account.0001_initial... OK
  Applying admin.0001_initial... OK
  Applying admin.0002_logentry_remove_auto_add... OK
  Applying contenttypes.0002_remove_content_type_name... OK
  Applying auth.0002_alter_permission_name_max_length... OK
  Applying auth.0003_alter_user_email_max_length... OK
  Applying auth.0004_alter_user_username_opts... OK
  Applying auth.0005_alter_user_last_login_null... OK
  Applying auth.0006_require_contenttypes_0002... OK
  Applying auth.0007_alter_validators_add_error_messages... OK
  Applying auth.0008_alter_user_username_max_length... OK
  Applying dashboard.0001_initial... OK
  Applying events.0001_initial... OK
  Applying sessions.0001_initial... OK
  Applying tagging.0001_initial... OK
  Applying tagging.0002_on_delete... OK
  Applying tags.0001_initial... OK
  Applying url_shortener.0001_initial... OK
```

`bin/service.sh start` is this script https://gist.github.com/RadoBuransky/9128967

Upon starting the service by `./dashboard.sh` I'm getting the below output
```
[ec2-user@ip-172-17-71-229 ~]$ ./dashboard.sh
starting carbon
2018/12/13 01:30:37 [WARNING] `common.log-level` and `common.logfile` is DEPRICATED. Use `logging` config section
starting graphite-web
starting dashboard
mkdir: cannot create directory ‘logs’: File exists
nohup: redirecting stderr to stdout
Starting dxedge-dashboard at port 9000...                             ERROR [Returned: 7]
ERROR [no response from server, timeout]
```

**Comment by deniszh** [comment]
@njchandu : this script is not part of Graphite installation and not required for its work, and we not able to provide help about it.
Graphite-web is working in port 8080, try to open that port in web browser. Is it working? If not - what in the logs?

**Comment by njchandu** [comment]
@deniszh I will re write the scripts and update it here.

**Comment by deniszh** [comment]
@njchandu : no need to touch script for now, just use browser...

**Comment by njchandu** [comment]
@deniszh where can I find the log file of graphite-web?

Thanks

**Comment by deniszh** [comment]
Usually it's in `/opt/graphite/webapp/log` but it's configurable.

**Comment by njchandu** [comment]
@deniszh 
I believe it is breaking here
```
+ echo 'Moving graphite storage folders'
+ mkdir /mnt/data/graphite
+ mv /opt/graphite/storage /mnt/data/graphite/storage
+ ln -s /mnt/data/graphite/storage /opt/graphite/storage
+ chown -R ec2-user:ec2-user /mnt/data/
+ cd /opt/graphite/webapp/graphite/
+ python ./manage.py syncdb --noinput
python: can't open file './manage.py': [Errno 2] No such file or directory
```
and the script is
```
function configure_graphite(){
  echo "Graphite/Carbon configuration"
  # Update links
  GRAPHITE_DATA=${DATA_MOUNT}/graphite
  if [ ! -d "${GRAPHITE_DATA}" ]; then
    echo "Moving graphite storage folders"
    mkdir ${GRAPHITE_DATA}
    mv /opt/graphite/storage ${GRAPHITE_DATA}/storage
  else
    echo "Graphite storage folders exist - removing"
    rm -r /opt/graphite/storage
  fi
  ln -s ${DATA_MOUNT}/graphite/storage /opt/graphite/storage
  chown -R ec2-user:ec2-user ${DATA_MOUNT}/

  cd /opt/graphite/webapp/graphite/
  python ./manage.py syncdb --noinput
  chown -R ec2-user:ec2-user ${DATA_MOUNT}/graphite/storage
  chown -R ec2-user:ec2-user /opt/graphite/storage
}
```

**Comment by deniszh** [comment]
Sorry, @njchandu, I'm really confused. 3 days ago you said that command was there and you ran it successfully - https://github.com/graphite-project/graphite-web/issues/2391#issuecomment-446809009
Even when you started this thread `django-admin.py` was there...
django-admin.py is part of Django installation if it's not there - then Django not installed properly.
Probably, you need to describe your setup and what you trying to achieve more thoroughly.

**Comment by njchandu** [comment]
@deniszh 
Apologies that happened because the set was happening from a root user. It works fine now. But, it doesn't load the data on the dashboard. The left navigation bar doesn't load. This is what I notice in the logs.

```
[16/Dec/2018 13:56:40] "GET / HTTP/1.1" 200 598
[16/Dec/2018 13:56:41] "GET /browser/header HTTP/1.1" 200 1167
[16/Dec/2018 13:56:41] "GET /composer? HTTP/1.1" 200 1643
[16/Dec/2018 13:56:41] "GET /static/js/browser.js HTTP/1.1" 404 94
[16/Dec/2018 13:56:41] "GET /static/js/ext/resources/css/ext-all.css HTTP/1.1" 404 113
[16/Dec/2018 13:56:41] "GET /static/js/ext/adapter/ext/ext-base.js HTTP/1.1" 404 111
[16/Dec/2018 13:56:41] "GET /static/js/ext/ext-all.js HTTP/1.1" 404 98
[16/Dec/2018 13:56:41] "GET /static/js/composer_widgets.js HTTP/1.1" 404 103
[16/Dec/2018 13:56:41] "GET /static/js/composer.js HTTP/1.1" 404 95
[16/Dec/2018 13:56:41] "GET /static/js/completer.js HTTP/1.1" 404 96
[16/Dec/2018 13:56:41] "GET /static/img/carbon-fiber.png HTTP/1.1" 404 101
[16/Dec/2018 13:56:41] "GET /static/img/graphite-logo.png HTTP/1.1" 404 102
[16/Dec/2018 13:56:41] "GET /static/js/browser.js HTTP/1.1" 404 94
[16/Dec/2018 13:56:42] "GET /static/js/composer_widgets.js HTTP/1.1" 404 103
[16/Dec/2018 13:56:42] "GET /static/js/composer.js HTTP/1.1" 404 95
[16/Dec/2018 13:56:42] "GET /static/js/completer.js HTTP/1.1" 404 96
```

**Comment by deniszh** [comment]
@njchandu:
you need to run collectstatic command `python django-admin.py collectstatic --settings=graphite.settings`

**Comment by njchandu** [comment]
Hi @piotr1212 , I'm getting the same error when starting the django server with `django-admin.py runserver --pythonpath /opt/graphite/webapp --settings graphite.settings 0.0.0.0:8080`

**Comment by piotr1212** [comment]
There are some issue with the way that the Django internal server handles static content, I can't remember the details, I think there was some security setting blocking this.

But why do you want to run the Django server? The recommended way is to use one of: apache+mod_wsgi, gunicorn, or uwsgi+nginx. as described in the documents https://graphite.readthedocs.io/en/latest/config-webapp.html

**Comment by piotr1212** [comment]
The original issue as reported seems to be solved, this is getting off-topic, closing. If you still have troubles please open a new issue.
